protocols/noise/src/lib.rs

```diff
@@ -157,42 +160,46 @@ where
 impl<T, C> InboundUpgrade<T> for NoiseConfig<IX, C>
 where
     NoiseConfig<IX, C>: UpgradeInfo,
-    T: AsyncRead + AsyncWrite + Send + 'static,
+    T: AsyncRead + AsyncWrite + Unpin + Send + 'static,
     C: Protocol<C> + AsRef<[u8]> + Zeroize + Send + 'static,
 {
     type Output = (RemoteIdentity<C>, NoiseOutput<Negotiated<T>>);
     type Error = NoiseError;
-    type Future = Handshake<Negotiated<T>, C>;
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Output, Self::Error>> + Send>>;
 
     fn upgrade_inbound(self, socket: Negotiated<T>, _: Self::Info) -> Self::Future {
-        let session = self.params.into_builder()
-            .local_private_key(self.dh_keys.secret().as_ref())
-            .build_responder()
-            .map_err(NoiseError::from);
-        Handshake::rt1_responder(socket, session,
-            self.dh_keys.into_identity(),
-            IdentityExchange::Mutual)
+        Box::pin(async move {
```

Comment:
Why is it necessary to wrap these blocks in another future that just waits on the inner future?

Reply:
We can't express the type of the inner future. Async functions are similar to closures in the way that the type they generate is totally opaque.

Reply:
It seems preferable to me though to keep `rt1_responder` & co return an opaque, boxed Future, as was done before with the [`Handshake`](https://github.com/libp2p/rust-libp2p/blob/master/protocols/noise/src/io/handshake.rs#L36) newtype, instead of declaring them as `async fn`s. Then the re-wrapping here in the declaration of the handshake patterns shouldn't be necessary, leaving the code and types much as before, with `type Future = Handshake<...>`.

Reply:
The old code is very hard to upgrade to new futures, because all the reading/writing helper functions provided by the `futures` library (such as `write_all` or `read_exact`) now hold the socket by reference instead of taking ownership of it.

On the other hand, the async/await syntax makes it possible to write code that is very easy to read (a linear list of instructions, as opposed to a state machine). This is one of the main reasons why we're upgrading to stable futures.

Reply:
> On the other hand, the async/await syntax makes it possible to write code that is very easy to read (a linear list of instructions, as opposed to a state machine). This is one of the main reasons why we're upgrading to stable futures.

I'm not suggesting not to use async/await for the implementation of `rt1_responder` et al, just to return an opaque `Handshake` future as before, to avoid the repeated re-wrapping in the definition of the handshake pattern configs. It may be my ignorance that I don't see why that wouldn't work and simplify things, maybe I have to try it myself to see what happens.

Reply:
See https://github.com/mxinden/rust-libp2p/pull/1.